pkg/services/ngalert/schedule/recording_rule.go

```diff
@@ -204,6 +237,7 @@ func (r *recordingRule) tryEvaluation(ctx context.Context, ev *Evaluation, logge
 	if err := eval.FindConditionError(result, ev.rule.Record.From); err != nil {
 		return fmt.Errorf("the query failed with an error: %w", err)
 	}
+	// TODO: This is missing dedicated logic for NoData. If NoData we can skip the write.
```

Comment:
Should we skip the write? I wonder if that should be decided by the writer instead.

Reply:
Putting this in the writer is a leaky abstraction. Check the logic down a few lines:
```
	frames, err := r.frameRef(ev.rule.Record.From, result)
	if err != nil {
		...
```
This is one of our two nodata states - frame missing. And, that code cannot go inside the writer, as the writer takes a single frame to begin with - we'd need to completely change the writer interface to accept the entire response list rather than a specific frame.

And also - this area needs a way to know that NoData happened, so we can update health correctly.